File: NeetCode/1-D_Dynamic_Programming/Tribonaci/Tribonaci.cpp
// Tribonaci.cpp
//
// Compile locally with:
// clang++ Tribonaci.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int tribonacci(int n)
    {
        if (n == 0)
        {
            return 0;
        }
        if (n == 1 || n == 2)
        {
            return 1;
        }

        vector<int> tri(n + 1, 0);
        tri[1] = 1;
        tri[2] = 1;
        for (int index = 3; index < n + 1; index++)
        {
            tri[index] = tri[index - 1] + tri[index - 2] + tri[index - 3];
        }
        return tri[n];
    }
};

int main()
{
    cout << "Running Tribonaci.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}

File: NeetCode/1-D_Dynamic_Programming/Tribonaci/tribonaci.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def tribonacci(self, n: int) -> int:
        if n == 0:
            return 0
        if n in (1, 2):
            return 1
        dp = [None for _ in range(n + 1)]
        dp[0] = 0
        dp[1] = 1
        dp[2] = 1
        for i in range(3, n + 1):
            dp[i] = dp[i - 1] + dp[i - 2] + dp[i - 3]

        return dp[n]

        pass


if __name__ == "__main__":
    sol = Solution()
    print(sol.tribonacci(3))
    print(sol.tribonacci(21))
    print("Running Solution...")
